src/main/java/org/opentripplanner/routing/edgetype/StreetEdge.java

```diff
@@ -679,8 +679,32 @@ private StateEditor doTraverse(State s0, RoutingRequest options, TraverseMode tr
 
             // Apply turn restrictions
             if (options.arriveBy && !canTurnOnto(backPSE, s0, backMode)) {
-                return null;
+                // A turn restriction exists that forbids this turn with the current traverseMode.
+                // If using a bike, or micromobility, try again while walking
+                if (traverseMode == TraverseMode.BICYCLE || traverseMode == TraverseMode.MICROMOBILITY) {
+                    return doTraverse(s0, options.bikeWalkingOptions, TraverseMode.WALK);
+                }
+                // After trying again with switching to walking, the backMode will not have changed. Therefore, this
+                // conditional makes an assumption that a transition from non-walking to walking occurs at the very end
+                // of the last state. In this case it might be possible to make said turn if an immediate transition to
+                // walking is assumed at the very end of the previous state.
+                else if (
+                    traverseMode == TraverseMode.WALK &&
+                        backMode != TraverseMode.WALK &&
+                        canTurnOnto(backPSE, s0, TraverseMode.WALK)
+                ) {
+                    // the turn is now possible with the assumption that a transition to walking occurred at the very
+                    // end of the last StreetEdge of the back state.
+                    // Do nothing here in order to continue the traversal and avoid returning null.
+                } else {
+                    return null;
+                }
             } else if (!options.arriveBy && !backPSE.canTurnOnto(this, s0, traverseMode)) {
+                // A turn restriction exists that forbids this turn with the current traverseMode.
+                // If using a bike, or micromobility, try again while walking
+                if (traverseMode == TraverseMode.BICYCLE || traverseMode == TraverseMode.MICROMOBILITY) {
+                    return doTraverse(s0, options.bikeWalkingOptions, TraverseMode.WALK);
+                }
```

Comment:
Because this conditional statement is used in two places, would it make sense to place the result of it in a boolean variable (e.g., `recurseTraversalWithWalk`), which gets reused in both places? I feel like that might increase the clarity of the code and prevent future mistakes where only one of the blocks might get updated. The whole block is repeated, but I don't think there's a good way to refactor the whole thing.